File: questions/01_2sum_indexes/code.js
var twoSum = function(nums, target) {
    const map = new Map();
    for (let i = 0; i < nums.length; i++) {
        let compliment = target - nums[i]; //determine the compliment we would need at current index

        if (map.has(compliment)) { //check our map for compliment
            return [map.get(compliment), i]; //return the compliment's index and current index in an array
        };

        map.set(nums[i],i); // iteratively set key and value pair in the map
    };
    return[]; //shell return as not to throw error, will never activate per guidelines
};

File: questions/01_2sum_indexes/code.go
package main

import (
    "fmt"
)

func twoSum(nums []int, target int) []int { // args are a slice of ints 'nums' and our target int, returns a slice of two ints (solution)
    hashMap := make(map[int]int) // initialize an empty hashmap key is an integer (a num from nums) and the value is defined as an integer(our index in nums).

    for i, num := range nums { // loop through nums.
        compliment := target - num // calculates compliment needed for current num in nums.

        if index, exists := hashMap[compliment]; exists { // exists is a boolean flag similar to 'truthy' concept in JS.
            return []int{index, i} // returns slice of integers: the current index of num in nums, and the saved value of the key compliment
        } 
        // else
        hashMap[num] = i // stores the current num in nums with the current index into the hashmap
    }
    return nil // edgecase for no solution
}